pipeline_dp/accumulator.py

```diff
@@ -0,0 +1,82 @@
+import abc
+import typing
+
+
+class Accumulator(abc.ABC):
+  """
+  Performs aggregations
+  """
+
+  @classmethod
+  def merge(cls, accumulators: typing.Iterable[
+    'Accumulator'] = None) -> 'Accumulator':
+    """
+    Merges the accumulators and creates an Accumulator
+    Args:
+      accumulators:
+
+    Returns: Accumulator instance with merged values
+    """
+    return cls(accumulators)
+
+  @abc.abstractmethod
+  def add_value(self, v):
+    """
+    Adds the value to each of the accumulator constituting the
+    CompoundAccumulator
+    Args:
+      v: value to be added
+
+    Returns:self
+    """
+    pass
+
+  @abc.abstractmethod
+  def add_accumulator(self, accumulator: 'Accumulator') -> 'Accumulator':
+    """
+      Merges the accumulator
+      The difference between this and the merge function is that here it
+      accepts a single accumulator instead of a list removing the overhead of
+      creating a list when merging.
+      Args:
+        accumulator:
+
+      Returns: self
+    """
+    pass
+
+  @abc.abstractmethod
+  def compute_metrics(self):
+    pass
+
+
+class CompoundAccumulator(Accumulator):
+  """
+  Performs aggregations for a compund accumulators
+  """
+
+  def __init__(self, accumulators: typing.Iterable['Accumulator'] = None):
+    self.accumulators = []
+    if accumulators:
+      # flatten the accumulators if the input is a CompundAccumulator
+      self.accumulators = [accumulator_expanded for accumulator in accumulators
+                           for accumulator_expanded in
+                           (accumulator.accumulators if isinstance(accumulator,
+                                                                   CompoundAccumulator)
+                            else [accumulator])]
+
+  def add_value(self, v):
+    # adds the value to each accumulator
+    for accumulator in self.accumulators:
+      accumulator.add_value(v)
+    return self
+
+  def add_accumulator(self, accumulator: 'CompoundAccumulator') -> \
+    'CompoundAccumulator':
+    # merges the accumulators of the CompoundAccumulators
+    self.accumulators.extend(accumulator.accumulators)
```

Comment:
Sry, I didn't explained that, the idea that we merge CompoundAcculators which contains accumulators of the same type and in the same order.
**add_accumulator()** will be called when we will have 1 accumulators, which have results from different worker machines.

E.g. let's assume that
```
compound_accumulator1.accumulators = [count_accumulator1, sum_accumulator1]
compound_accumulator2.accumulators = [count_accumulator2, sum_accumulator2]
```
then `compound_accumulator1.add_accumulator(compound_accumulator2) `should do
```
count_accumulator1.add_accumulator(count_accumulator2)
sum_accumulator1.add_accumulator(sum_accumulator2)
```

i.e. `add_accumulator(acc)` accumulates data from `acc`.

Reply:
Thanks for explaining! I misunderstood the first time!